Performance.md

```diff
@@ -0,0 +1,290 @@
+This is to understand how to emit features like `const enum`s and `namespace`s.
+But needing to check other files to emit of an arbitrary file can make emit slower.
+
+The need for features that need non-local information is somewhat rare - regular `enum`s can be used in place of `const enum`s, and modules can be used instead of `namespace`s.
+For that reason, TypeScript provides the `isolatedModules` flag to warn when using them.
+Using `isolatedModules` means that your codebase is safe for tools to use TypeScript APIs like `transpileModule` or alternative compilers like Babel.
+
+As an example of this in action, ts-loader provides
+
+* the `transpileOnly` flag to use `transpileModule`.
+* the `useBabel` flag
+
+<!-- TODO -->
+
+# Investigating Issues
+
+There are certain ways to get hints of what might be going wrong.
+
+## `extendedDiagnostics`
+
+You can run TypeScript with `--extendedDiagnostics` to get a printout of where the compiler is spending its time.
+
+```
+Files:                         6
+Lines:                     24906
+Nodes:                    112200
+Identifiers:               41097
+Symbols:                   27972
+Types:                      8298
+Memory used:              77984K
+Assignability cache size:  33123
+Identity cache size:           2
+Subtype cache size:            0
+I/O Read time:             0.01s
+Parse time:                0.44s
+Program time:              0.45s
+Bind time:                 0.21s
+Check time:                1.07s
+transformTime time:        0.01s
+commentTime time:          0.00s
+I/O Write time:            0.00s
+printTime time:            0.01s
+Emit time:                 0.01s
+Total time:                1.75s
+```
+
+The most relevant information for most users is:
+
+Field   | Meaning
+--------|---------
+`Files` | the number of files that the program is including (use `listFiles` to see what they are).
+`I/O Read time` | time spent reading from the file system - this includes traversing `include`'d folders.
+`Parse time` | time spent scanning and parsing the program
+`Program time` | combined time spent performing reading from the file system, scanning and parsing the program, and other calculation of the program graph. These steps are intermingled and combined here because files need to be resolved and loaded once they're included via `import`s and `export`s.
+`Bind time` | Time spent building up various semantic information that is local to a single file.
+`Check time` | Time spent type-checking the program.
+`transformTime time` | Time spent rewriting TypeScript ASTs (trees that represent source files) into forms that work in older runtimes.
+`commentTime` | Time spent calculating comments in output files.
+`I/O Write time` | Time spent writing/updating files on disk.
+`printTime` | Time spent calculating the string representation of an output file.
+
+<!-- TODO: is this important? Looks identical to printTime
+
+`Emit time` | ...
```

Comment:
I've seen it be a little slower than printTime but I don't know what it excludes.

Reply:
emitTime = printTime  + tranform time (i think)